src/coreclr/tools/Common/Compiler/DisplayNameHelpers.cs

```diff
@@ -116,7 +117,58 @@ public static string GetDisplayName(this EventPseudoDesc @event)
 
         public static string GetDisplayName(this TypeDesc type)
         {
-            return Formatter.Instance.FormatName(type, FormatOptions.NamespaceQualify);
+            var sb = new StringBuilder();
+            if (type == null)
+                return sb.ToString();
+
+            Stack<TypeDesc> genericArguments = null;
+            switch (type)
+            {
+                // TODO: Linker handles here ArrayType and GenericInstanceType
+                default:
+                    string simpleName = Formatter.Instance.FormatName(type, FormatOptions.NamespaceQualify);
+                    if (type.HasInstantiation)
+                    {
+                        int genericParametersCount = type.Instantiation.Length;
+                        int declaringTypeGenericParametersCount = 0;
+                        if (type is DefType defType)
+                            declaringTypeGenericParametersCount = defType.ContainingType?.Instantiation.Length ?? 0;
+
+                        if (genericParametersCount > declaringTypeGenericParametersCount)
+                        {
+                            if (genericArguments?.Count > 0)
+                                PrependGenericArguments(genericArguments, genericParametersCount - declaringTypeGenericParametersCount, sb);
+                            else
+                                PrependGenericParameters(type.Instantiation, sb);
+
+                            int explicitArityIndex = simpleName.IndexOf('`');
+                            simpleName = explicitArityIndex != -1 ? simpleName.Substring(0, explicitArityIndex) : simpleName;
```

Comment:
What happens for nested types? I think there will be backticks on multiple segments potentially.